### ISSUE TITLE
protobuf 3.0.0 tests compile hangs infinitely when compiled with gcc 6.1.0 on MSYS2 (cmake build)

### ISSUE THREAD

**Issue by emmenlau** [google/protobuf#1940]
I'm compiling protobuf with cmake on recent MSYS2, which now ships gcc 6.1.0. The compile seems to work fine as long as I disable tests. With tests enabled, gcc takes very long on some files, and eventually hangs infinitely on

```
[ 79%] Building CXX object CMakeFiles/tests.dir/src/google/protobuf/util/internal/protostream_objectsource_test.cc.obj
```

The cmake command I use is (with TESTS="OFF" it works, with TESTS="ON" it hangs):

```
  MSYS2_ARG_CONV_EXCL="-DCMAKE_INSTALL_PREFIX=" \
  ${MINGW_PREFIX}/bin/cmake ./cmake \
      -Wno-dev \
      -G"MSYS Makefiles" \
      -DCMAKE_INSTALL_PREFIX="${MINGW_PREFIX}" \
      -DCMAKE_BUILD_TYPE="Release" \
      -DCMAKE_SKIP_RPATH="ON" \
      -Dprotobuf_MSVC_STATIC_RUNTIME="ON" \
      -DBUILD_SHARED_LIBS="ON" \
      -Dprotobuf_BUILD_TESTS="ON" \
      -Dprotobuf_WITH_ZLIB="ON"
```

**Comment by emmenlau** [comment]
FYI: protobuf 3.0.0-beta2 still compiled fine with tests (same build instructions).

**Comment by xfxyjwf** [comment]
Seems a compiler bug when compiling some complicated C++ templates. I suggest try a different g++ version.

**Comment by emmenlau** [comment]
I don't have much of a choice there. I can report the problem to MSYS2 development team, in the hope that they will upgrade gcc soon again. But I'm pretty sure gcc 6.2 is not out yet, and I don't think MSYS2 will downgrade gcc easily.

**Comment by jagij** [comment]
This seems to be fixed in some version of gcc > 6.1.1: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71507
Don't know when it will be available though..

**Comment by emmenlau** [comment]
Cool thanks for checking!

**Comment by emmenlau** [comment]
Fixed with gcc 6.2.0 on newest MSYS2.
